File: Hello LeetCode/LeetCodeBook-Java/src/com/algorithm/链表/_141_环形链表.java
package com.algorithm.链表;

/**
 * https://leetcode-cn.com/problems/linked-list-cycle/
 */
public class _141_环形链表 {

    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;

        ListNode slowNode = head;

        ListNode fastNode = head.next;

        // 这里条件判断不能用 ||，要用 &&,比如链表 [1,2]
        while (fastNode != null || fastNode.next != null) {
            if (slowNode == fastNode) return true;
            fastNode = fastNode.next.next;
            slowNode = slowNode.next;
        }
        return false;
    }
}

File: Hello LeetCode/LeetCodeBook-Swift/Leetcode-Book/Leetcode-Book/LinkedList/_141_环形链表.swift
//
//  _141_环形链表.swift
//  Leetcode-Book
//
//  Created by gaoliang on 2022/1/18.
//

import Foundation

// https://leetcode-cn.com/problems/linked-list-cycle/
class _141_环形链表_Solution {
    
    /// 快慢指针
    func hasCycle(_ head: ListNode?) -> Bool {
        var fast = head?.next
        var slow = head
        
        while fast != nil && slow != nil {
            if fast == slow {
                return true
            } else {
                fast = fast?.next?.next
                slow = slow?.next
            }
        }
        return false
    }
    
    /// 通过哈希表
    func _hasCycle(_ head: ListNode?) -> Bool {
        var hashMap = [ListNode:ListNode]()
        
        var node = head
        
        while node != nil {
            /// 如果有
            if hashMap[node!] != nil {
                return true
            } else {
                hashMap[node!] = node
            }
            node = node?.next
        }
        
        return false
    }
}
